membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -258,30 +266,36 @@ class AccountController(
           subscription <- SimpleEitherT(
             services.subscriptionService
               .current(contact)
-              .map(subs => subscriptionSelector(subscriptionName, s"the sfUser $contact", subs)),
+              .map(subs => subscriptionSelector(memsub.Subscription.Name(subscriptionName), s"the sfUser $contact", subs)),
           )
-          contributionPlan <- SimpleEitherT.fromEither(subscription.plan match {
-            case p if p.product == Contribution => Right(p)
+          catalog <- SimpleEitherT.rightT(services.futureCatalog)
+          contributionPlan = subscription.plan(catalog)
+          _ <- SimpleEitherT.fromEither(contributionPlan.product(catalog) match {
+            case Contribution => Right(())
             case nc => Left(s"$subscriptionName plan is not a contribution: " + nc)
           })
-          billingPeriod = contributionPlan.charges.billingPeriod.asInstanceOf[RecurringPeriod]
-          applyFromDate = contributionPlan.chargedThrough.getOrElse(contributionPlan.start)
-          currency = contributionPlan.charges.price.prices.head.currency
+          billingPeriod <- SimpleEitherT.fromEither(contributionPlan.billingPeriod.toEither)
+          recurringPeriod <- SimpleEitherT.fromEither(billingPeriod match {
+            case period: RecurringPeriod => Right(period)
+            case period: BillingPeriod.OneOffPeriod => Left(s"period $period was not recurring for contribution update")
+          })
+          applyFromDate = contributionPlan.chargedThroughDate.getOrElse(contributionPlan.start)
+          currency = contributionPlan.chargesPrice.prices.head.currency
           currencyGlyph = currency.glyph
-          oldPrice = contributionPlan.charges.price.prices.head.amount
+          oldPrice = contributionPlan.chargesPrice.prices.head.amount
           reasonForChange =
             s"User updated contribution via self-service MMA. Amount changed from $currencyGlyph$oldPrice to $currencyGlyph$newPrice effective from $applyFromDate"
           result <- SimpleEitherT(
             services.zuoraRestService.updateChargeAmount(
               subscription.name,
-              contributionPlan.charges.subRatePlanChargeId,
+              contributionPlan.ratePlanCharges.head.id,
```

Comment:
Do we know that ratePlanCharges has a head value?

Reply:
yes good point, in reality there is always (exactly) one charge in the rateplan which would be the contribution.
If a broken contribution subscription came in then they would not even be able to see it in manage, let alone update the charge amount.
In the old system it would fail to deserialise the broken contribution, now it would only notice the issue when it comes to use it at this point.
It does raise a good point (along with the fact that your paper subscription found an issue).  Before if there was an issue with the data it was discovered by the deserialiser and the subscription was usually silently discarded.  Now if there is an issue, it will trip up when it tries to use it.  I wonder how many broken subscriptions (or incomplete traits etc) we have that slip under the radar.